expanded weights: embedding faster rule

samdow opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70141
* #73696
* #73694
* __->__ #73693

Differential Revision: [D34719301](https://our.internmc.facebook.com/intern/diff/D34719301)

Commit:
expanded weights: embedding faster rule

[ghstack-poisoned]

Commit:
Update on "expanded weights: embedding faster rule"




[ghstack-poisoned]

Commit:
Update on "expanded weights: embedding faster rule"




[ghstack-poisoned]

Commit:
Update on "expanded weights: embedding faster rule"




[ghstack-poisoned]

Commit:
Update on "expanded weights: embedding faster rule"




[ghstack-poisoned]

Commit:
Update on "expanded weights: embedding faster rule"




[ghstack-poisoned]

samdow commented on test/test_expanded_weights.py:
```diff
@@ -187,10 +187,14 @@ def test_unsupported_expand_weights(self, device, dtype, op):
         for sample_input in supported_inputs(op, sample_inputs):
+            if op.name == "nn.functional.embedding":  # embedding flips its argument order for autograd tests
+                sample_input = SampleInput(sample_input.args[0], args=(sample_input.input,), kwargs=sample_input.kwargs)
+                if 'max_norm' in sample_input.kwargs:  # in place update makes it difficult even with copying
```
The fact that this worked on the last commit is a little suss to me because it suggests that the test that was failing (linux-xenial-cuda11.3-py3.7-gcc7) is failing because it's not copying. I'll try to make a smaller repro

zou3519 replied:
@samdow is this still failing? I'd like to understand what the problem is before this goes in (it sounds concerning)

samdow commented:
@samdow has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34719301).

Commit:
Update on "expanded weights: embedding faster rule"


Differential Revision: [D34719301](https://our.internmc.facebook.com/intern/diff/D34719301)

[ghstack-poisoned]

samdow commented:
@samdow has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34719301).

Commit:
Update on "expanded weights: embedding faster rule"


Differential Revision: [D34719301](https://our.internmc.facebook.com/intern/diff/D34719301)

[ghstack-poisoned]

Commit:
Update on "expanded weights: embedding faster rule"


Differential Revision: [D34719301](https://our.internmc.facebook.com/intern/diff/D34719301)

[ghstack-poisoned]

Commit:
Update on "expanded weights: embedding faster rule"


Differential Revision: [D34719301](https://our.internmc.facebook.com/intern/diff/D34719301)

[ghstack-poisoned]

Commit:
Update on "expanded weights: embedding faster rule"


Differential Revision: [D34719301](https://our.internmc.facebook.com/intern/diff/D34719301)

[ghstack-poisoned]

Commit:
Update on "expanded weights: embedding faster rule"


Differential Revision: [D34719301](https://our.internmc.facebook.com/intern/diff/D34719301)

[ghstack-poisoned]

zou3519 commented on torch/nn/utils/_expanded_weights/embedding_expanded_weights.py:
```diff
@@ -0,0 +1,54 @@
+class EmbeddingPerSampleGrad(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, kwarg_names, _, *expanded_args_and_kwargs):
+        expanded_args, expanded_kwargs = standard_kwargs(kwarg_names, expanded_args_and_kwargs)
```
Nit: if I am understanding this correctly, the operators that don't standardize are the ones that go through the PyTorch dispatcher first. Is that right?

https://github.com/pytorch/pytorch/blob/fb47cff737ec5aa5f3451db5c877410e0ae18141/torch/nn/utils/_expanded_weights/expanded_weights_utils.py#L5-L6

samdow replied:
I think that it's the ones where the `handle_torch_function` is generated in the Python/C++ binding layer. I can't find the exact generation for this but I think it will look something like [this] (https://github.com/pytorch/pytorch/blob/40d1f77384672337bd7e734e32cb5fad298959bd/tools/autograd/templates/python_nn_functions.cpp#L46) where it just passes the args and kwargs through as PyObjects

So I think it hits the Python/C++ binding layer but not the PyTorch dispatcher (but I might be missing something)

zou3519 reviewed (approved):
LGTM, but we should understand why the test you pointed out above was failing (if we haven't yet). It might be an indication that something deeper is problematic (or might just be an unrelated testing quirk)

Commit:
Update on "expanded weights: embedding faster rule"


Differential Revision: [D34719301](https://our.internmc.facebook.com/intern/diff/D34719301)

[ghstack-poisoned]

samdow commented on test/test_expanded_weights.py:
```diff
@@ -187,19 +187,17 @@ def test_unsupported_expand_weights(self, device, dtype, op):
+            normal_result = run_op(op, sample_input.input, *sample_input.args, **sample_input.kwargs)
             self.assertEqual(expanded_weight_result, normal_result)
 
-    def test_expanded_weight_error(self, device):
-        batch_size = 3
-        sample_input = make_tensor((batch_size, 4), dtype=torch.float32, device=device, requires_grad=True)
-        sample_weight = make_tensor((4), dtype=torch.float32, device=device, requires_grad=True)
-        with self.assertRaisesRegex(RuntimeError, r"Expanded Weights encountered but cannot handle function"):
-            torch.add(sample_input, ExpandedWeight(sample_weight, batch_size))
```
Thanks! I think this got deleted as I was tinkering around with testing the embedding issue. I'll replace it

Commit:
Update on "expanded weights: embedding faster rule"


Differential Revision: [D34719301](https://our.internmc.facebook.com/intern/diff/D34719301)

[ghstack-poisoned]

Commit:
Update on "expanded weights: embedding faster rule"


Differential Revision: [D34719301](https://our.internmc.facebook.com/intern/diff/D34719301)

[ghstack-poisoned]

Commit:
Update on "expanded weights: embedding faster rule"


Differential Revision: [D34719301](https://our.internmc.facebook.com/intern/diff/D34719301)

[ghstack-poisoned]

Commit:
Update on "expanded weights: embedding faster rule"


Differential Revision: [D34719301](https://our.internmc.facebook.com/intern/diff/D34719301)

[ghstack-poisoned]

Commit:
Update on "expanded weights: embedding faster rule"


Differential Revision: [D34719301](https://our.internmc.facebook.com/intern/diff/D34719301)

[ghstack-poisoned]

Commit:
Update on "expanded weights: embedding faster rule"


Differential Revision: [D34719301](https://our.internmc.facebook.com/intern/diff/D34719301)

[ghstack-poisoned]

Commit:
Update on "expanded weights: embedding faster rule"


Differential Revision: [D34719301](https://our.internmc.facebook.com/intern/diff/D34719301)

[ghstack-poisoned]

Commit:
Update on "expanded weights: embedding faster rule"


Differential Revision: [D34719301](https://our.internmc.facebook.com/intern/diff/D34719301)

[ghstack-poisoned]

Commit:
Update on "expanded weights: embedding faster rule"


Differential Revision: [D34719301](https://our.internmc.facebook.com/intern/diff/D34719301)

[ghstack-poisoned]

samdow commented:
@pytorchbot merge this please
